{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "690e282b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils import MeltingTemp\n",
    "from Bio.SeqUtils import gc_fraction\n",
    "import pandas as pd\n",
    "import re\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "58810460",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_codon_usage_table(file_name) :\n",
    "\n",
    "    # Read the Excel file into a DataFrame\n",
    "    df = pd.read_excel(file_name)\n",
    "\n",
    "    # Create a dictionary from the Excel columns\n",
    "    amino_acids = df[\"Amino acid\"].tolist()\n",
    "    codons_1 = df.groupby(\"Amino acid\")[\"Codon 1\"].apply(list).to_dict()\n",
    "    codons_2 = df.groupby(\"Amino acid\")[\"Codon 2\"].apply(list).to_dict()\n",
    "    \n",
    "    return amino_acids, codons_1, codons_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5a8dcf87",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_plasmid(file_name) :\n",
    "    \n",
    "    with open(file_name, 'r') as plasmid_file:\n",
    "        plasmid_lines = plasmid_file.readlines()\n",
    "\n",
    "    plasmid_seq = Seq(plasmid_lines[0].strip())                     #Return\n",
    "    start_gene_index = int(plasmid_lines[1]) - 1                    #Return\n",
    "    end_gene_index = int(plasmid_lines[2]) - 1\n",
    "    gene_seq = plasmid_seq[start_gene_index:(end_gene_index + 1)]   #Return\n",
    "    protein_seq = gene_seq.translate()                              #Return\n",
    "    \n",
    "    return plasmid_seq, start_gene_index, gene_seq, protein_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ba7e0327",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_mutations(file_name) :\n",
    "\n",
    "    # Create an empty list to store the values\n",
    "    mutations_list = []\n",
    "\n",
    "    # Count the number of mutations\n",
    "    mutations_count = 0\n",
    "\n",
    "    # Open the file for reading\n",
    "    with open(file_name, 'r') as mutations_file:\n",
    "        # Read each line and append it to the list\n",
    "        for line in mutations_file:\n",
    "            mutations_list.append(line.strip())  # strip() removes any leading/trailing whitespace    #Return\n",
    "            mutations_count += 1                                                                      #Return\n",
    "    \n",
    "    return mutations_list, mutations_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "36c56047",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify and separate components in mutations\n",
    "def separate_mutation(mutation) :\n",
    "    \n",
    "    print('Mutation now is ', mutation)\n",
    "\n",
    "    # Define a regular expression pattern to match a letter followed by a number and then another letter\n",
    "    pattern = r'([A-Za-z])(\\d{1,4})([A-Za-z])'\n",
    "\n",
    "    # Use the search function to find the match in the input string\n",
    "    match = re.search(pattern, mutation)\n",
    "\n",
    "    # Check if a match was found\n",
    "    if match:\n",
    "        # Extract the individual components into variables\n",
    "        original_AA = match.group(1)\n",
    "        position = match.group(2)\n",
    "        new_AA = match.group(3)\n",
    "    else:\n",
    "        print(\"Could not find correct mutation pattern.\")\n",
    "        sys.exit()\n",
    "\n",
    "    return original_AA, position, new_AA\n",
    "\n",
    "def separate_all_mutations(mutations_list) :\n",
    "    \n",
    "    mutations_separated = [[],[],[]]\n",
    "\n",
    "    for mutation in mutations_list :\n",
    "        separated_mutation = separate_mutation(mutation)\n",
    "        mutations_separated[0].append(separated_mutation[0])      #Return\n",
    "        mutations_separated[1].append(int(separated_mutation[1])) #Return\n",
    "        mutations_separated[2].append(separated_mutation[2])      #Return\n",
    "        \n",
    "    return mutations_separated[0], mutations_separated[1], mutations_separated[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ef8b1555",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find position of mutation in the gene:\n",
    "\n",
    "def make_mutations(mutations_count, mutation_position_AAs, new_AAs, gene_seq, codons_1, start_gene_index, plasmid_seq) :\n",
    "\n",
    "    original_codons = []\n",
    "    new_codons = []\n",
    "    codons_start_index_plasmid = []\n",
    "    new_plasmid_seqs = []\n",
    "\n",
    "    for mutation in range(mutations_count) :\n",
    "        codon_start_gene = ((mutation_position_AAs[mutation] - 1) * 3)\n",
    "        codon_end_gene = codon_start_gene + 3\n",
    "        original_codon = gene_seq[codon_start_gene:codon_end_gene]\n",
    "        print('Original codon is ', original_codon)\n",
    "        original_codons.append(original_codon)                     #Return\n",
    "        new_codon = Seq((str(codons_1[new_AAs[mutation]])).strip(\"[]'\"))\n",
    "        print('New codon is ', new_codon)\n",
    "        new_codons.append(new_codon)                               #Return\n",
    "        codon_start_plasmid = start_gene_index + codon_start_gene\n",
    "        codons_start_index_plasmid.append(codon_start_plasmid)     #Return\n",
    "        codon_end_plasmid = start_gene_index + codon_end_gene\n",
    "        new_plasmid_seq = plasmid_seq[:codon_start_plasmid] + new_codon + plasmid_seq[codon_end_plasmid:]\n",
    "        new_plasmid_seqs.append(new_plasmid_seq)                   #Return\n",
    "    \n",
    "    return original_codons, new_codons, new_plasmid_seqs, codons_start_index_plasmid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "65eb1561",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_mutation_positions(original_codon, new_codon, codon_start_plasmid) :\n",
    "    found_first_position = False\n",
    "    found_last_position = False\n",
    "    for position in range(0,3) :\n",
    "        print('Base in original codon is ', original_codon[position])\n",
    "        print('Base in new codon is ', new_codon[position])\n",
    "        if original_codon[position] != new_codon[position] and found_first_position == False :\n",
    "            first_mutation_position = codon_start_plasmid + position\n",
    "            found_first_position = True\n",
    "            print('First base mutation found at position ', first_mutation_position)\n",
    "            continue\n",
    "        if original_codon[position] != new_codon[position] and found_first_position == True :\n",
    "            last_mutation_position = codon_start_plasmid + position\n",
    "            found_last_position = True\n",
    "            print('Last base mutation found at position ', last_mutation_position)\n",
    "    if found_last_position == False :\n",
    "        last_mutation_position = first_mutation_position\n",
    "        print('Last base mutation not found, so first base mutation will be used, which is in index ', last_mutation_position)\n",
    "        \n",
    "    return first_mutation_position, last_mutation_position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b73296dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_overlap_flanking_position_left(new_plasmid_seq, start_search_position_left) :\n",
    "    current_position = start_search_position_left - 1\n",
    "    base_at_current_position = new_plasmid_seq[current_position]\n",
    "    print('Starting to look at position ', current_position, 'with base ', base_at_current_position)\n",
    "    \n",
    "    while base_at_current_position not in ['T', 't'] :\n",
    "        current_position -= 1                          #Return\n",
    "        if current_position == -1 :\n",
    "            print('\\nError: T could not be found at the left side of the mutations.')\n",
    "            sys.exit()\n",
    "        base_at_current_position = new_plasmid_seq[current_position]\n",
    "        print('Current index is ', current_position, ' whith base ', base_at_current_position)\n",
    "    \n",
    "    print('\\nT flanking on the left side of the mutations was found at index ', current_position)\n",
    "        \n",
    "    return current_position\n",
    "\n",
    "\n",
    "def find_overlap_flanking_position_right(new_plasmid_seq, start_search_position_right) :\n",
    "    current_position = start_search_position_right + 1\n",
    "    base_at_current_position = new_plasmid_seq[current_position]\n",
    "    last_plasmid_index = len(new_plasmid_seq) - 1\n",
    "    print('Starting to look at position ', current_position, 'with base ', base_at_current_position)\n",
    "    \n",
    "    while base_at_current_position not in ['A', 'a'] :\n",
    "        current_position += 1                          #Return\n",
    "        if current_position == last_plasmid_index :\n",
    "            print('\\nError: A could not be found at the right side of the mutations.')\n",
    "            sys.exit()\n",
    "        base_at_current_position = new_plasmid_seq[current_position]\n",
    "        print('Current index is ', current_position, ' whith base ', base_at_current_position)\n",
    "    \n",
    "    print('\\nA flanking on the right side of the mutations was found at index ', current_position)\n",
    "        \n",
    "    return current_position\n",
    "    \n",
    "\n",
    "def find_overlap_flanking_positions(start_search_position_left, start_search_position_right, new_plasmid_seq) :\n",
    "    \n",
    "    overlap_flanking_position_left = find_overlap_flanking_position_left(new_plasmid_seq, start_search_position_left)    #Return\n",
    "    overlap_flanking_position_right = find_overlap_flanking_position_right(new_plasmid_seq, start_search_position_right) #Return\n",
    "    \n",
    "    return overlap_flanking_position_left, overlap_flanking_position_right"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5f250308",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_overlap_end_position_left(overlap_flanking_position_left, plasmid) :\n",
    "    search_range_end_overlap = range((overlap_flanking_position_left - 5), (overlap_flanking_position_left - 10), -1)\n",
    "    overlap_end_condition = ['A', 'a']\n",
    "    \n",
    "    print('\\nCurrently looking for the A to make an overlap flanking the left side of the mutation.')\n",
    "    print('The index range to look for the last overlap base is', search_range_end_overlap)\n",
    "\n",
    "    for position in search_range_end_overlap:\n",
    "        print('Current base is', plasmid[position])\n",
    "        if plasmid[position] in overlap_end_condition:\n",
    "            overlap_end_position_left = position                        #Return\n",
    "            print('Found end position at', overlap_end_position_left)\n",
    "            break\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "    return overlap_end_position_left\n",
    "\n",
    "    \n",
    "def find_overlap_end_position_right(overlap_flanking_position_right, plasmid) :\n",
    "    search_range_end_overlap = range((overlap_flanking_position_right + 5), (overlap_flanking_position_right + 10))\n",
    "    overlap_end_condition = ['T', 't']\n",
    "    \n",
    "    print('\\nCurrently looking for the T to make an overlap flanking the left side of the mutation.')\n",
    "    print('The index range to look for the last overlap base is', search_range_end_overlap)\n",
    "\n",
    "    for position in search_range_end_overlap:\n",
    "        print('Current base is', plasmid[position])\n",
    "        if plasmid[position] in overlap_end_condition:\n",
    "            overlap_end_position_right = position                        #Return\n",
    "            print('Found end position at', overlap_end_position_right)\n",
    "            break\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "    return overlap_end_position_right\n",
    "    \n",
    "\n",
    "def find_overlap_end_positions(plasmid, overlap_flanking_position_left, overlap_flanking_position_right) :\n",
    "    \n",
    "    overlap_end_position_left = find_overlap_end_position_left(overlap_flanking_position_left, plasmid)     #Return\n",
    "                                                                                                                                                                              \n",
    "    overlap_end_position_right = find_overlap_end_position_right(overlap_flanking_position_right, plasmid)  #Return\n",
    "                                                                                                                                                                                  \n",
    "    return overlap_end_position_left, overlap_end_position_right"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c0cdafb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_overlap_seq(plasmid, overlap_flanking_position_left, overlap_end_position_left, \n",
    "                     overlap_flanking_position_right, overlap_end_position_right) :\n",
    "\n",
    "    overlap_left_fw = plasmid[overlap_end_position_left:(overlap_flanking_position_left + 1)]\n",
    "    overlap_left_rv = overlap_left_fw.reverse_complement()\n",
    "    overlap_right_fw = plasmid[overlap_flanking_position_right:(overlap_end_position_right + 1)]\n",
    "    overlap_right_rv = overlap_right_fw.reverse_complement()\n",
    "\n",
    "    print('Left forward overlapping sequence is: ', overlap_left_fw)\n",
    "    print('Left reverse overlapping sequence is: ', overlap_left_rv)\n",
    "    print('Right forward overlapping sequence is: ', overlap_right_fw)\n",
    "    print('Right reverse overlapping sequence is: ', overlap_right_rv)\n",
    "    \n",
    "    return overlap_left_fw, overlap_left_rv, overlap_right_fw, overlap_right_rv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "63e76931",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_single_non_overlap(plasmid, start_position, non_overlap_search_range, primer_direction, t_anneal):\n",
    "    print('\\nCurrently calculating the non-overlapping region for the primer.')\n",
    "\n",
    "    start_position = start_position + primer_direction\n",
    "    print('\\nStart search position is: ', start_position, 'with base', plasmid[start_position])\n",
    "\n",
    "    for non_overlap_length in non_overlap_search_range:\n",
    "        current_position = start_position + non_overlap_length\n",
    "        print('\\nCurrent position is: ', current_position)\n",
    "        print('Base in this position is: ', plasmid[current_position])\n",
    "\n",
    "        if current_position > start_position:\n",
    "            seq = plasmid[start_position:(current_position + 1)]\n",
    "        else:\n",
    "            seq = plasmid[current_position:(start_position + 1)]\n",
    "\n",
    "        tm = MeltingTemp.Tm_NN(seq)\n",
    "\n",
    "        if (tm >= t_anneal and seq[-1] in ['C', 'c', 'G', 'g']) or (tm >= t_anneal and seq[0] in ['C', 'c', 'G', 'g']):\n",
    "            break\n",
    "\n",
    "    else:\n",
    "        print('\\nError: could not resolve non-overlapping Tm for the primer.')\n",
    "        sys.exit()\n",
    "        \n",
    "    if tm > (t_anneal + 5) :\n",
    "        print('\\nError: Oh, no! The Tm for one of the non-overlapping regions is too high! **cry**')\n",
    "        sys.exit()\n",
    "        \n",
    "    if len(seq) < 5 :\n",
    "        print('\\nError: Hm... The length of this non-overlapping region is too short. **sad**')\n",
    "        sys.exit()\n",
    "\n",
    "    print('\\nThe non-overlapping sequence of the primer of this insert is', seq)\n",
    "    print('and the Tm is', tm)\n",
    "\n",
    "    return seq, tm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "69df42f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_all_non_overlaps(plasmid, last_mutation_position, overlap_end_position_left, first_mutation_position, \n",
    "                          overlap_end_position_right, t_anneal) :\n",
    "    \n",
    "    non_overlap_search_range_fwd = range(10, 60)\n",
    "    non_overlap_search_range_rv = range(-10, -60, -1)\n",
    "    \n",
    "    non_overlap_left_fw = make_single_non_overlap(plasmid, last_mutation_position, non_overlap_search_range_fwd, 1, t_anneal)\n",
    "    non_overlap_left_fw_seq = non_overlap_left_fw[0]\n",
    "    tm_left_fw = non_overlap_left_fw[1]\n",
    "    \n",
    "    non_overlap_left_rv = make_single_non_overlap(plasmid, overlap_end_position_left, non_overlap_search_range_rv, -1, t_anneal)\n",
    "    non_overlap_left_rv_seq = non_overlap_left_rv[0].reverse_complement()\n",
    "    tm_left_rv = non_overlap_left_rv[1]\n",
    "    \n",
    "    non_overlap_right_fw = make_single_non_overlap(plasmid, overlap_end_position_right, non_overlap_search_range_fwd, 1, t_anneal)\n",
    "    non_overlap_right_fw_seq = non_overlap_right_fw[0]\n",
    "    tm_right_fw = non_overlap_right_fw[1]\n",
    "    \n",
    "    non_overlap_right_rv = make_single_non_overlap(plasmid, first_mutation_position, non_overlap_search_range_rv, -1, t_anneal)\n",
    "    non_overlap_right_rv_seq = non_overlap_right_rv[0].reverse_complement()\n",
    "    tm_right_rv = non_overlap_right_rv[1]\n",
    "       \n",
    "    return non_overlap_left_fw_seq, tm_left_fw, non_overlap_left_rv_seq, tm_left_rv, non_overlap_right_fw_seq, tm_right_fw, non_overlap_right_rv_seq, tm_right_rv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "dc1d4016",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_inter_sequence(plasmid, overlap_flanking_position_left, last_mutation_position, first_mutation_position, \n",
    "                        overlap_flanking_position_right) :\n",
    "\n",
    "    inter_seq_left_fw = plasmid[(overlap_flanking_position_left + 1):(last_mutation_position + 1)]\n",
    "    inter_seq_right_rv = plasmid[first_mutation_position:overlap_flanking_position_right].reverse_complement()\n",
    "    \n",
    "    print('Left forward primer inter-sequence is ', inter_seq_left_fw)\n",
    "    print('Right reverse primer inter-sequence is ', inter_seq_right_rv)\n",
    "\n",
    "    return inter_seq_left_fw, inter_seq_right_rv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "41fc09d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fuse(overlap_left_fw, overlap_left_rv, overlap_right_fw, overlap_right_rv, non_overlap_left_fw, non_overlap_left_rv, \n",
    "         non_overlap_right_fw, non_overlap_right_rv, inter_seq_left_fw, inter_seq_right_rv):\n",
    "        \n",
    "    primer_left_fw = overlap_left_fw[:-1] + 'U' + inter_seq_left_fw + non_overlap_left_fw\n",
    "    print('The final left forward primer is ', primer_left_fw)\n",
    "    primer_left_rv = overlap_left_rv[:-1] + 'U' + non_overlap_left_rv\n",
    "    print('The final left reverse primer is ', primer_left_rv)\n",
    "    primer_right_fw = overlap_right_fw[:-1] + 'U' + non_overlap_right_fw\n",
    "    print('The final right forward primer is ', primer_right_fw)\n",
    "    primer_right_rv = overlap_right_rv[:-1] + 'U' + inter_seq_right_rv + non_overlap_right_rv\n",
    "    print('The final right reverse primer is ', primer_right_rv)\n",
    "    \n",
    "    return primer_left_fw, primer_left_rv, primer_right_fw, primer_right_rv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d42fa853",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_gc_content(primer_sequence) :\n",
    "    \n",
    "    gc_content = gc_fraction(primer_sequence) * 100  # as a percentage\n",
    "    \n",
    "    return gc_content\n",
    "\n",
    "\n",
    "def check_gc_content_all_primers(primer_left_fw, primer_left_rv, primer_right_fw, primer_right_rv) :\n",
    "    \n",
    "    gc_content_primer_left_fw = check_gc_content(primer_left_fw)\n",
    "    gc_content_primer_left_rv = check_gc_content(primer_left_rv)\n",
    "    gc_content_primer_right_fw = check_gc_content(primer_right_fw)\n",
    "    gc_content_primer_right_rv = check_gc_content(primer_right_rv)\n",
    "\n",
    "    print('\\nGC content of primer_left_fw is ', gc_content_primer_left_fw)\n",
    "    print('\\nGC content of primer_left_rv is ', gc_content_primer_left_rv)\n",
    "    print('\\nGC content of primer_right_fw is ', gc_content_primer_right_fw)\n",
    "    print('\\nGC content of primer_right_rv is ', gc_content_primer_right_rv)\n",
    "    \n",
    "    return gc_content_primer_left_fw, gc_content_primer_left_rv, gc_content_primer_right_fw, gc_content_primer_right_rv "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "67a0d594-5b0d-49ef-9f4e-b6e11460e141",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_primer(primer, gc_content,) :\n",
    "    primer_unusable = False\n",
    "    \n",
    "    if gc_content < 30 or gc_content > 70 :\n",
    "        primer_unusable = True\n",
    "        print('\\nThe GC content of at least one primer in this pair is at least 20% off the ideal 50% content, so this primer pair will be considered unusable.')\n",
    "         \n",
    "    if len(primer) < 15 or len(primer) > 60 :\n",
    "        primer_unusable = True\n",
    "        print('\\nThe length of at least one primer in this pair is too short (<15 bp) or too long (>60 bp), so this primer pair will be considered unusable.')\n",
    "        \n",
    "    return primer_unusable\n",
    "\n",
    "def check_tm_pair(tm_fw, tm_rv, t_anneal) :\n",
    "    primers_unusable = False\n",
    "\n",
    "    if abs(tm_fw - tm_rv) > 5 :\n",
    "        print('\\nThe Tm difference between primer pair is too big (>5), so this primer pair will be considered unusable.')\n",
    "        primers_unusable = True\n",
    "\n",
    "    if abs(tm_fw - t_anneal) > 5 or abs(tm_rv - t_anneal) > 5 :\n",
    "        print('\\nThe Tm of at least one primer in this pair is too far from the annealing temperature, so this primer pair will be considered unusable.')\n",
    "        primers_unusable = True\n",
    "    \n",
    "    return primers_unusable\n",
    "\n",
    "def calculate_homopolymeric_penalty(primer) :\n",
    "    count = 0\n",
    "    sum = 0\n",
    "    length = len(primer)\n",
    "    \n",
    "    for n in range(1, length) :\n",
    "        if primer[n - 1] == primer[n] :\n",
    "            count += 1\n",
    "        else :\n",
    "            if count >= 4 :\n",
    "                penalty = count - 4\n",
    "                sum += penalty\n",
    "        count = 0\n",
    "\n",
    "    score = sum * 10\n",
    "\n",
    "    print('Homopolymeric penalty for this primer is ', score)\n",
    "    \n",
    "    return score\n",
    "\n",
    "def score_primers(t_anneal, primer_left_fw, primer_left_rv, primer_right_fw, primer_right_rv, tm_left_fw, tm_left_rv, tm_right_fw, \n",
    "                  tm_right_rv, gc_content_primer_left_fw, gc_content_primer_left_rv, gc_content_primer_right_fw, gc_content_primer_right_rv) :\n",
    "\n",
    "    primers_unusable_left, primers_unusable_right = False, False\n",
    "    \n",
    "    primer_left_fw_unusable = check_primer(primer_left_fw, gc_content_primer_left_fw)\n",
    "    primer_left_rv_unusable = check_primer(primer_left_rv, gc_content_primer_left_rv)\n",
    "    primer_right_fw_unusable = check_primer(primer_right_fw, gc_content_primer_right_fw)\n",
    "    primer_right_rv_unusable = check_primer(primer_right_rv, gc_content_primer_right_rv)\n",
    "\n",
    "    if (primer_left_fw_unusable == True) or (primer_left_rv_unusable == True) :\n",
    "        primers_unusable_left = True\n",
    "    else :\n",
    "        primer_pair_left_unusable = check_tm_pair(tm_left_fw, tm_left_rv, t_anneal)\n",
    "        if primer_pair_left_unusable == True :\n",
    "            primers_unusable_left = True\n",
    "\n",
    "    if (primer_right_fw_unusable == True) or (primer_right_rv_unusable == True) :\n",
    "        primers_unusable_right = True\n",
    "    else :\n",
    "        primer_pair_right_unusable = check_tm_pair(tm_right_fw, tm_right_rv, t_anneal)\n",
    "        if primer_pair_right_unusable == True :\n",
    "            primers_unusable_right = True\n",
    "    \n",
    "    if (primers_unusable_left == False) and (primers_unusable_right == True) :\n",
    "        primer_fw = primer_left_fw\n",
    "        primer_rv = primer_left_rv\n",
    "        tm_fw = tm_left_fw\n",
    "        tm_rv = tm_left_rv\n",
    "    elif (primers_unusable_right == False) and (primers_unusable_left == True) :\n",
    "        primer_fw = primer_right_fw\n",
    "        primer_rv = primer_right_rv\n",
    "        tm_fw = tm_right_fw\n",
    "        tm_rv = tm_right_rv\n",
    "    elif (primers_unusable_right == True) and (primers_unusable_left == True) :\n",
    "        primer_fw = str('Fail')\n",
    "        primer_rv = str('Fail')\n",
    "        print('Error: All designed primers for this mutant violated at least one critical rule for primer design. Hence, no primer will be proposed.')\n",
    "    else:\n",
    "        score_length_left = len(primer_left_fw) + len(primer_left_rv)\n",
    "        score_length_right = len(primer_right_fw) + len(primer_right_rv)\n",
    "        score_gc_content_left = abs(gc_content_primer_left_fw - 50) + abs(gc_content_primer_left_rv - 50)\n",
    "        score_gc_content_right = abs(gc_content_primer_right_fw - 50) + abs(gc_content_primer_right_rv - 50)\n",
    "        score_homopolymeric_penalty_left = calculate_homopolymeric_penalty(primer_left_fw) + calculate_homopolymeric_penalty(primer_left_rv)\n",
    "        score_homopolymeric_penalty_right = calculate_homopolymeric_penalty(primer_right_fw) + calculate_homopolymeric_penalty(primer_right_rv)\n",
    "        \n",
    "        total_score_left = score_length_left + score_gc_content_left + score_homopolymeric_penalty_left\n",
    "        total_score_right = score_length_right + score_gc_content_right + score_homopolymeric_penalty_right\n",
    "\n",
    "        print('Total score left is ', total_score_left)\n",
    "        print('Total score right is ', total_score_right)\n",
    "    \n",
    "        if total_score_left > total_score_right :\n",
    "            primer_fw = primer_right_fw\n",
    "            primer_rv = primer_right_rv\n",
    "            tm_fw = tm_right_fw\n",
    "            tm_rv = tm_right_rv\n",
    "        else :\n",
    "            primer_fw = primer_left_fw\n",
    "            primer_rv = primer_left_rv\n",
    "            tm_fw = tm_left_fw\n",
    "            tm_rv = tm_left_rv\n",
    "\n",
    "    return primer_fw, tm_fw, primer_rv, tm_rv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8fc11fa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def design_primer(original_codon, new_codon, codon_start_plasmid, mutation, plasmid) :\n",
    "\n",
    "    t_anneal = 56\n",
    "    \n",
    "    print('\\nCurrently determining the position of the first and last base change in the mutation ', mutation)\n",
    "    first_mutation_position, last_mutation_position = find_mutation_positions(original_codon, new_codon, codon_start_plasmid)\n",
    "    \n",
    "    start_search_position_left, start_search_position_right = first_mutation_position, last_mutation_position\n",
    "    \n",
    "    overlap_flanking_position_left, overlap_flanking_position_right = False, False\n",
    "    overlap_end_position_left, overlap_end_position_right = False, False\n",
    "    \n",
    "    while (overlap_end_position_left == False) or (overlap_end_position_right == False) :\n",
    "        if (overlap_end_position_left == False) and (overlap_flanking_position_left != False):\n",
    "            start_search_position_left = overlap_flanking_position_left\n",
    "        if (overlap_end_position_right == False) and (overlap_flanking_position_right != False):\n",
    "            start_search_position_right = overlap_flanking_position_right\n",
    "            \n",
    "        print('\\nCurrently looking for the first base to make an overlap flanking the mutation ', mutation)\n",
    "        overlap_flanking_position_left, overlap_flanking_position_right = find_overlap_flanking_positions(start_search_position_left,\n",
    "                                                                                                                                                           start_search_position_right,\n",
    "                                                                                                                                                           plasmid)\n",
    "    \n",
    "        print('\\nCurrently looking for the last base to make an overlap flanking the mutation ', mutation)\n",
    "        overlap_end_position_left, overlap_end_position_right = find_overlap_end_positions(plasmid, \n",
    "                                                                                           overlap_flanking_position_left, \n",
    "                                                                                           overlap_flanking_position_right)\n",
    "    \n",
    "    print('\\nMaking the overlapping sequence for the mutation ', mutation)\n",
    "    overlap_left_fw, overlap_left_rv, overlap_right_fw, overlap_right_rv = make_overlap_seq(plasmid, \n",
    "                                                                                            overlap_flanking_position_left, \n",
    "                                                                                            overlap_end_position_left, \n",
    "                                                                                            overlap_flanking_position_right,\n",
    "                                                                                            overlap_end_position_right)\n",
    "    \n",
    "    print('\\nMaking the non-overlapping sequence for the mutation ', mutation)\n",
    "    non_overlap_left_fw, tm_left_fw, non_overlap_left_rv, tm_left_rv, non_overlap_right_fw, tm_right_fw, non_overlap_right_rv, tm_right_rv = make_all_non_overlaps(plasmid, \n",
    "                                                                                                                                                                   last_mutation_position,\n",
    "                                                                                                                                                                   overlap_end_position_left,\n",
    "                                                                                                                                                                   first_mutation_position,\n",
    "                                                                                                                                                                   overlap_end_position_right,\n",
    "                                                                                                                                                                   t_anneal)\n",
    "    print('\\nMaking the inter-sequence for the mutation ', mutation)\n",
    "    inter_seq_left_fw, inter_seq_right_rv = make_inter_sequence(plasmid, overlap_flanking_position_left,\n",
    "                                                                last_mutation_position, first_mutation_position,\n",
    "                                                                overlap_flanking_position_right)\n",
    "    \n",
    "    print('\\nReplacing T with U in the overlapping region and combining with the non-overlapping region to finalize the primers for the variant ', mutation)\n",
    "    primer_left_fw, primer_left_rv, primer_right_fw, primer_right_rv = fuse(overlap_left_fw, overlap_left_rv, \n",
    "                                                                            overlap_right_fw, overlap_right_rv, \n",
    "                                                                            non_overlap_left_fw, non_overlap_left_rv,\n",
    "                                                                            non_overlap_right_fw, non_overlap_right_rv,\n",
    "                                                                            inter_seq_left_fw, inter_seq_right_rv)\n",
    "\n",
    "    print('\\nCalculating GC content of designed primers.')\n",
    "    \n",
    "    gc_content_primer_left_fw, gc_content_primer_left_rv, gc_content_primer_right_fw, gc_content_primer_right_rv = check_gc_content_all_primers(primer_left_fw, \n",
    "                                                                                                                                                primer_left_rv, \n",
    "                                                                                                                                                primer_right_fw, \n",
    "                                                                                                                                                primer_right_rv)\n",
    "\n",
    "    primer_fw, tm_fw, primer_rv, tm_rv = score_primers(t_anneal, \n",
    "                                         primer_left_fw, primer_left_rv, primer_right_fw, primer_right_rv, \n",
    "                                         tm_left_fw, tm_left_rv, tm_right_fw, tm_right_rv, \n",
    "                                         gc_content_primer_left_fw, gc_content_primer_left_rv, gc_content_primer_right_fw, gc_content_primer_right_rv)\n",
    "    \n",
    "    return primer_fw, tm_fw, primer_rv, tm_rv\n",
    "        \n",
    "\n",
    "def bulk_primer_design(mutations_count, original_codons, new_codons, codons_start_index_plasmid, mutations_list,\n",
    "                       new_plasmid_seqs) :\n",
    "    \n",
    "    primers_list = []\n",
    "    \n",
    "    for plasmid in range(mutations_count) :\n",
    "        primer_fw, tm_fw, primer_rv, tm_rv = design_primer(original_codons[plasmid], new_codons[plasmid],\n",
    "                                                           codons_start_index_plasmid[plasmid], mutations_list[plasmid],\n",
    "                                                           new_plasmid_seqs[plasmid])\n",
    "        \n",
    "        primers_list.append([primer_fw, tm_fw, primer_rv, tm_rv])  #Return\n",
    "        \n",
    "    return primers_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4ba7ee46",
   "metadata": {},
   "outputs": [],
   "source": [
    "def transpose_primers(primers_list, mutations_count) :\n",
    "    transposed_primers_list = [[primers_list[j][i] for j in range(mutations_count)] for i in range(len(primers_list[0]))]\n",
    "    \n",
    "    return transposed_primers_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "43049eb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def export_primers(transposed_primers_list, mutations_list, mutations_count) :\n",
    "    \n",
    "    final_primers_seq_list = []\n",
    "    final_primers_names_list = []\n",
    "    final_primers_number_list = []\n",
    "    final_primers_tm_list = []\n",
    "\n",
    "    for number in range(mutations_count):\n",
    "        final_primers_seq_list.append(transposed_primers_list[0][number])\n",
    "        final_primers_seq_list.append(transposed_primers_list[2][number])\n",
    "\n",
    "        final_primers_names_list.append(mutations_list[number] + ' fw')\n",
    "        final_primers_names_list.append(mutations_list[number] + ' rv')\n",
    "\n",
    "        final_primers_tm_list.append(transposed_primers_list[1][number])\n",
    "        final_primers_tm_list.append(transposed_primers_list[3][number])\n",
    "\n",
    "        for n in range(1, 3):\n",
    "            final_primers_number_list.append(n)\n",
    "            \n",
    "    df = pd.DataFrame({'Name': final_primers_names_list, 'Number': final_primers_number_list, 'Short name': None, 'Sequence': final_primers_seq_list, 'Tm': final_primers_tm_list})\n",
    "    df.to_excel(\"primers_list.xlsx\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "fc66e339",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main() :\n",
    "    amino_acids, codons_1, codons_2 = load_codon_usage_table('codon_usage_frequency_escherichia_coli_b.xlsx')\n",
    "    plasmid_seq, start_gene_index, gene_seq, protein_seq = load_plasmid('plasmid.txt')\n",
    "    mutations_list, mutations_count = load_mutations('mutations.txt')\n",
    "    \n",
    "    original_AAs, mutation_position_AAs, new_AAs = separate_all_mutations(mutations_list)\n",
    "    \n",
    "    original_codons, new_codons, new_plasmid_seqs, codons_start_index_plasmid = make_mutations(mutations_count, \n",
    "                                                                                               mutation_position_AAs, \n",
    "                                                                                               new_AAs, gene_seq, codons_1,\n",
    "                                                                                               start_gene_index, plasmid_seq)\n",
    "    \n",
    "    primers_list = bulk_primer_design(mutations_count, original_codons, new_codons, codons_start_index_plasmid,\n",
    "                                      mutations_list, new_plasmid_seqs)\n",
    "                                      \n",
    "    transposed_primers_list = transpose_primers(primers_list, mutations_count)\n",
    "    \n",
    "    export_primers(transposed_primers_list, mutations_list, mutations_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "c215091f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mutation now is  S10A\n",
      "Mutation now is  R13K\n",
      "Mutation now is  A15H\n",
      "Original codon is  TCG\n",
      "New codon is  GCG\n",
      "Original codon is  CGG\n",
      "New codon is  AAA\n",
      "Original codon is  GCT\n",
      "New codon is  CAT\n",
      "\n",
      "Currently determining the position of the first and last base change in the mutation  S10A\n",
      "Base in original codon is  T\n",
      "Base in new codon is  G\n",
      "First base mutation found at position  231\n",
      "Base in original codon is  C\n",
      "Base in new codon is  C\n",
      "Base in original codon is  G\n",
      "Base in new codon is  G\n",
      "Last base mutation not found, so first base mutation will be used, which is in index  231\n",
      "\n",
      "Currently looking for the first base to make an overlap flanking the mutation  S10A\n",
      "Starting to look at position  230 with base  C\n",
      "Current index is  229  whith base  C\n",
      "Current index is  228  whith base  G\n",
      "Current index is  227  whith base  T\n",
      "\n",
      "T flanking on the left side of the mutations was found at index  227\n",
      "Starting to look at position  232 with base  C\n",
      "Current index is  233  whith base  G\n",
      "Current index is  234  whith base  G\n",
      "Current index is  235  whith base  C\n",
      "Current index is  236  whith base  T\n",
      "Current index is  237  whith base  G\n",
      "Current index is  238  whith base  C\n",
      "Current index is  239  whith base  T\n",
      "Current index is  240  whith base  C\n",
      "Current index is  241  whith base  G\n",
      "Current index is  242  whith base  G\n",
      "Current index is  243  whith base  A\n",
      "\n",
      "A flanking on the right side of the mutations was found at index  243\n",
      "\n",
      "Currently looking for the last base to make an overlap flanking the mutation  S10A\n",
      "\n",
      "Currently looking for the A to make an overlap flanking the left side of the mutation.\n",
      "The index range to look for the last overlap base is range(222, 217, -1)\n",
      "Current base is G\n",
      "Current base is C\n",
      "Current base is T\n",
      "Current base is A\n",
      "Found end position at 219\n",
      "\n",
      "Currently looking for the T to make an overlap flanking the left side of the mutation.\n",
      "The index range to look for the last overlap base is range(248, 253)\n",
      "Current base is T\n",
      "Found end position at 248\n",
      "\n",
      "Making the overlapping sequence for the mutation  S10A\n",
      "Left forward overlapping sequence is:  ATCGTGATT\n",
      "Left reverse overlapping sequence is:  AATCACGAT\n",
      "Right forward overlapping sequence is:  ACGGCT\n",
      "Right reverse overlapping sequence is:  AGCCGT\n",
      "\n",
      "Making the non-overlapping sequence for the mutation  S10A\n",
      "\n",
      "Currently calculating the non-overlapping region for the primer.\n",
      "\n",
      "Start search position is:  232 with base C\n",
      "\n",
      "Current position is:  242\n",
      "Base in this position is:  G\n",
      "\n",
      "Current position is:  243\n",
      "Base in this position is:  A\n",
      "\n",
      "Current position is:  244\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  245\n",
      "Base in this position is:  G\n",
      "\n",
      "Current position is:  246\n",
      "Base in this position is:  G\n",
      "\n",
      "Current position is:  247\n",
      "Base in this position is:  C\n",
      "\n",
      "The non-overlapping sequence of the primer of this insert is CGGCTGCTCGGACGGC\n",
      "and the Tm is 58.04384864846389\n",
      "\n",
      "Currently calculating the non-overlapping region for the primer.\n",
      "\n",
      "Start search position is:  218 with base G\n",
      "\n",
      "Current position is:  208\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  207\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  206\n",
      "Base in this position is:  G\n",
      "\n",
      "Current position is:  205\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  204\n",
      "Base in this position is:  A\n",
      "\n",
      "Current position is:  203\n",
      "Base in this position is:  g\n",
      "\n",
      "Current position is:  202\n",
      "Base in this position is:  t\n",
      "\n",
      "Current position is:  201\n",
      "Base in this position is:  a\n",
      "\n",
      "Current position is:  200\n",
      "Base in this position is:  t\n",
      "\n",
      "Current position is:  199\n",
      "Base in this position is:  a\n",
      "\n",
      "Current position is:  198\n",
      "Base in this position is:  c\n",
      "\n",
      "Current position is:  197\n",
      "Base in this position is:  c\n",
      "\n",
      "Current position is:  196\n",
      "Base in this position is:  g\n",
      "\n",
      "Current position is:  195\n",
      "Base in this position is:  a\n",
      "\n",
      "Current position is:  194\n",
      "Base in this position is:  c\n",
      "\n",
      "Current position is:  193\n",
      "Base in this position is:  g\n",
      "\n",
      "The non-overlapping sequence of the primer of this insert is gcagccatatgATGTCAACTCCTTCG\n",
      "and the Tm is 58.13697145041624\n",
      "\n",
      "Currently calculating the non-overlapping region for the primer.\n",
      "\n",
      "Start search position is:  249 with base G\n",
      "\n",
      "Current position is:  259\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  260\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  261\n",
      "Base in this position is:  A\n",
      "\n",
      "Current position is:  262\n",
      "Base in this position is:  A\n",
      "\n",
      "Current position is:  263\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  264\n",
      "Base in this position is:  G\n",
      "\n",
      "Current position is:  265\n",
      "Base in this position is:  G\n",
      "\n",
      "Current position is:  266\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  267\n",
      "Base in this position is:  G\n",
      "\n",
      "Current position is:  268\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  269\n",
      "Base in this position is:  G\n",
      "\n",
      "The non-overlapping sequence of the primer of this insert is GTGGGTTCCTTTAACGGTGCG\n",
      "and the Tm is 56.3263213661059\n",
      "\n",
      "Currently calculating the non-overlapping region for the primer.\n",
      "\n",
      "Start search position is:  230 with base C\n",
      "\n",
      "Current position is:  220\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  219\n",
      "Base in this position is:  A\n",
      "\n",
      "Current position is:  218\n",
      "Base in this position is:  G\n",
      "\n",
      "Current position is:  217\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  216\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  215\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  214\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  213\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  212\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  211\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  210\n",
      "Base in this position is:  A\n",
      "\n",
      "Current position is:  209\n",
      "Base in this position is:  A\n",
      "\n",
      "Current position is:  208\n",
      "Base in this position is:  C\n",
      "\n",
      "The non-overlapping sequence of the primer of this insert is CAACTCCTTCGATCGTGATTGCC\n",
      "and the Tm is 56.292662752663944\n",
      "\n",
      "Making the inter-sequence for the mutation  S10A\n",
      "Left forward primer inter-sequence is  GCCG\n",
      "Right reverse primer inter-sequence is  CCGAGCAGCCGC\n",
      "\n",
      "Replacing T with U in the overlapping region and combining with the non-overlapping region to finalize the primers for the variant  S10A\n",
      "The final left forward primer is  ATCGTGATUGCCGCGGCTGCTCGGACGGC\n",
      "The final left reverse primer is  AATCACGAUCGAAGGAGTTGACATcatatggctgc\n",
      "The final right forward primer is  ACGGCUGTGGGTTCCTTTAACGGTGCG\n",
      "The final right reverse primer is  AGCCGUCCGAGCAGCCGCGGCAATCACGATCGAAGGAGTTG\n",
      "\n",
      "Calculating GC content of designed primers.\n",
      "\n",
      "GC content of primer_left_fw is  71.42857142857143\n",
      "\n",
      "GC content of primer_left_rv is  47.05882352941176\n",
      "\n",
      "GC content of primer_right_fw is  61.53846153846154\n",
      "\n",
      "GC content of primer_right_rv is  65.0\n",
      "\n",
      "The GC content of at least one primer in this pair is at least 20% off the ideal 50% content, so this primer pair will be considered unusable.\n",
      "\n",
      "Currently determining the position of the first and last base change in the mutation  R13K\n",
      "Base in original codon is  C\n",
      "Base in new codon is  A\n",
      "First base mutation found at position  240\n",
      "Base in original codon is  G\n",
      "Base in new codon is  A\n",
      "Last base mutation found at position  241\n",
      "Base in original codon is  G\n",
      "Base in new codon is  A\n",
      "Last base mutation found at position  242\n",
      "\n",
      "Currently looking for the first base to make an overlap flanking the mutation  R13K\n",
      "Starting to look at position  239 with base  T\n",
      "\n",
      "T flanking on the left side of the mutations was found at index  239\n",
      "Starting to look at position  243 with base  A\n",
      "\n",
      "A flanking on the right side of the mutations was found at index  243\n",
      "\n",
      "Currently looking for the last base to make an overlap flanking the mutation  R13K\n",
      "\n",
      "Currently looking for the A to make an overlap flanking the left side of the mutation.\n",
      "The index range to look for the last overlap base is range(234, 229, -1)\n",
      "Current base is G\n",
      "Current base is G\n",
      "Current base is C\n",
      "Current base is T\n",
      "Current base is C\n",
      "\n",
      "Currently looking for the T to make an overlap flanking the left side of the mutation.\n",
      "The index range to look for the last overlap base is range(248, 253)\n",
      "Current base is T\n",
      "Found end position at 248\n",
      "\n",
      "Currently looking for the first base to make an overlap flanking the mutation  R13K\n",
      "Starting to look at position  238 with base  C\n",
      "Current index is  237  whith base  G\n",
      "Current index is  236  whith base  T\n",
      "\n",
      "T flanking on the left side of the mutations was found at index  236\n",
      "Starting to look at position  243 with base  A\n",
      "\n",
      "A flanking on the right side of the mutations was found at index  243\n",
      "\n",
      "Currently looking for the last base to make an overlap flanking the mutation  R13K\n",
      "\n",
      "Currently looking for the A to make an overlap flanking the left side of the mutation.\n",
      "The index range to look for the last overlap base is range(231, 226, -1)\n",
      "Current base is T\n",
      "Current base is C\n",
      "Current base is C\n",
      "Current base is G\n",
      "Current base is T\n",
      "\n",
      "Currently looking for the T to make an overlap flanking the left side of the mutation.\n",
      "The index range to look for the last overlap base is range(248, 253)\n",
      "Current base is T\n",
      "Found end position at 248\n",
      "\n",
      "Currently looking for the first base to make an overlap flanking the mutation  R13K\n",
      "Starting to look at position  235 with base  C\n",
      "Current index is  234  whith base  G\n",
      "Current index is  233  whith base  G\n",
      "Current index is  232  whith base  C\n",
      "Current index is  231  whith base  T\n",
      "\n",
      "T flanking on the left side of the mutations was found at index  231\n",
      "Starting to look at position  243 with base  A\n",
      "\n",
      "A flanking on the right side of the mutations was found at index  243\n",
      "\n",
      "Currently looking for the last base to make an overlap flanking the mutation  R13K\n",
      "\n",
      "Currently looking for the A to make an overlap flanking the left side of the mutation.\n",
      "The index range to look for the last overlap base is range(226, 221, -1)\n",
      "Current base is T\n",
      "Current base is A\n",
      "Found end position at 225\n",
      "\n",
      "Currently looking for the T to make an overlap flanking the left side of the mutation.\n",
      "The index range to look for the last overlap base is range(248, 253)\n",
      "Current base is T\n",
      "Found end position at 248\n",
      "\n",
      "Making the overlapping sequence for the mutation  R13K\n",
      "Left forward overlapping sequence is:  ATTGCCT\n",
      "Left reverse overlapping sequence is:  AGGCAAT\n",
      "Right forward overlapping sequence is:  ACGGCT\n",
      "Right reverse overlapping sequence is:  AGCCGT\n",
      "\n",
      "Making the non-overlapping sequence for the mutation  R13K\n",
      "\n",
      "Currently calculating the non-overlapping region for the primer.\n",
      "\n",
      "Start search position is:  243 with base A\n",
      "\n",
      "Current position is:  253\n",
      "Base in this position is:  G\n",
      "\n",
      "Current position is:  254\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  255\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  256\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  257\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  258\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  259\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  260\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  261\n",
      "Base in this position is:  A\n",
      "\n",
      "Current position is:  262\n",
      "Base in this position is:  A\n",
      "\n",
      "Current position is:  263\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  264\n",
      "Base in this position is:  G\n",
      "\n",
      "The non-overlapping sequence of the primer of this insert is ACGGCTGTGGGTTCCTTTAACG\n",
      "and the Tm is 57.45081965199222\n",
      "\n",
      "Currently calculating the non-overlapping region for the primer.\n",
      "\n",
      "Start search position is:  224 with base G\n",
      "\n",
      "Current position is:  214\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  213\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  212\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  211\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  210\n",
      "Base in this position is:  A\n",
      "\n",
      "Current position is:  209\n",
      "Base in this position is:  A\n",
      "\n",
      "Current position is:  208\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  207\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  206\n",
      "Base in this position is:  G\n",
      "\n",
      "Current position is:  205\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  204\n",
      "Base in this position is:  A\n",
      "\n",
      "Current position is:  203\n",
      "Base in this position is:  g\n",
      "\n",
      "Current position is:  202\n",
      "Base in this position is:  t\n",
      "\n",
      "Current position is:  201\n",
      "Base in this position is:  a\n",
      "\n",
      "Current position is:  200\n",
      "Base in this position is:  t\n",
      "\n",
      "Current position is:  199\n",
      "Base in this position is:  a\n",
      "\n",
      "Current position is:  198\n",
      "Base in this position is:  c\n",
      "\n",
      "The non-overlapping sequence of the primer of this insert is catatgATGTCAACTCCTTCGATCGTG\n",
      "and the Tm is 56.05645960154874\n",
      "\n",
      "Currently calculating the non-overlapping region for the primer.\n",
      "\n",
      "Start search position is:  249 with base G\n",
      "\n",
      "Current position is:  259\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  260\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  261\n",
      "Base in this position is:  A\n",
      "\n",
      "Current position is:  262\n",
      "Base in this position is:  A\n",
      "\n",
      "Current position is:  263\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  264\n",
      "Base in this position is:  G\n",
      "\n",
      "Current position is:  265\n",
      "Base in this position is:  G\n",
      "\n",
      "Current position is:  266\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  267\n",
      "Base in this position is:  G\n",
      "\n",
      "Current position is:  268\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  269\n",
      "Base in this position is:  G\n",
      "\n",
      "The non-overlapping sequence of the primer of this insert is GTGGGTTCCTTTAACGGTGCG\n",
      "and the Tm is 56.3263213661059\n",
      "\n",
      "Currently calculating the non-overlapping region for the primer.\n",
      "\n",
      "Start search position is:  239 with base T\n",
      "\n",
      "Current position is:  229\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  228\n",
      "Base in this position is:  G\n",
      "\n",
      "Current position is:  227\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  226\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  225\n",
      "Base in this position is:  A\n",
      "\n",
      "Current position is:  224\n",
      "Base in this position is:  G\n",
      "\n",
      "Current position is:  223\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  222\n",
      "Base in this position is:  G\n",
      "\n",
      "Current position is:  221\n",
      "Base in this position is:  C\n",
      "\n",
      "The non-overlapping sequence of the primer of this insert is CGTGATTGCCTCGGCTGCT\n",
      "and the Tm is 57.58770027191105\n",
      "\n",
      "Making the inter-sequence for the mutation  R13K\n",
      "Left forward primer inter-sequence is  CGGCTGCTAAA\n",
      "Right reverse primer inter-sequence is  TTT\n",
      "\n",
      "Replacing T with U in the overlapping region and combining with the non-overlapping region to finalize the primers for the variant  R13K\n",
      "The final left forward primer is  ATTGCCUCGGCTGCTAAAACGGCTGTGGGTTCCTTTAACG\n",
      "The final left reverse primer is  AGGCAAUCACGATCGAAGGAGTTGACATcatatg\n",
      "The final right forward primer is  ACGGCUGTGGGTTCCTTTAACGGTGCG\n",
      "The final right reverse primer is  AGCCGUTTTAGCAGCCGAGGCAATCACG\n",
      "\n",
      "Calculating GC content of designed primers.\n",
      "\n",
      "GC content of primer_left_fw is  53.84615384615385\n",
      "\n",
      "GC content of primer_left_rv is  45.45454545454545\n",
      "\n",
      "GC content of primer_right_fw is  61.53846153846154\n",
      "\n",
      "GC content of primer_right_rv is  59.25925925925925\n",
      "Homopolymeric penalty for this primer is  0\n",
      "Homopolymeric penalty for this primer is  0\n",
      "Homopolymeric penalty for this primer is  0\n",
      "Homopolymeric penalty for this primer is  0\n",
      "Total score left is  82.3916083916084\n",
      "Total score right is  75.79772079772079\n",
      "\n",
      "Currently determining the position of the first and last base change in the mutation  A15H\n",
      "Base in original codon is  G\n",
      "Base in new codon is  C\n",
      "First base mutation found at position  246\n",
      "Base in original codon is  C\n",
      "Base in new codon is  A\n",
      "Last base mutation found at position  247\n",
      "Base in original codon is  T\n",
      "Base in new codon is  T\n",
      "\n",
      "Currently looking for the first base to make an overlap flanking the mutation  A15H\n",
      "Starting to look at position  245 with base  G\n",
      "Current index is  244  whith base  C\n",
      "Current index is  243  whith base  A\n",
      "Current index is  242  whith base  G\n",
      "Current index is  241  whith base  G\n",
      "Current index is  240  whith base  C\n",
      "Current index is  239  whith base  T\n",
      "\n",
      "T flanking on the left side of the mutations was found at index  239\n",
      "Starting to look at position  248 with base  T\n",
      "Current index is  249  whith base  G\n",
      "Current index is  250  whith base  T\n",
      "Current index is  251  whith base  G\n",
      "Current index is  252  whith base  G\n",
      "Current index is  253  whith base  G\n",
      "Current index is  254  whith base  T\n",
      "Current index is  255  whith base  T\n",
      "Current index is  256  whith base  C\n",
      "Current index is  257  whith base  C\n",
      "Current index is  258  whith base  T\n",
      "Current index is  259  whith base  T\n",
      "Current index is  260  whith base  T\n",
      "Current index is  261  whith base  A\n",
      "\n",
      "A flanking on the right side of the mutations was found at index  261\n",
      "\n",
      "Currently looking for the last base to make an overlap flanking the mutation  A15H\n",
      "\n",
      "Currently looking for the A to make an overlap flanking the left side of the mutation.\n",
      "The index range to look for the last overlap base is range(234, 229, -1)\n",
      "Current base is G\n",
      "Current base is G\n",
      "Current base is C\n",
      "Current base is T\n",
      "Current base is C\n",
      "\n",
      "Currently looking for the T to make an overlap flanking the left side of the mutation.\n",
      "The index range to look for the last overlap base is range(266, 271)\n",
      "Current base is T\n",
      "Found end position at 266\n",
      "\n",
      "Currently looking for the first base to make an overlap flanking the mutation  A15H\n",
      "Starting to look at position  238 with base  C\n",
      "Current index is  237  whith base  G\n",
      "Current index is  236  whith base  T\n",
      "\n",
      "T flanking on the left side of the mutations was found at index  236\n",
      "Starting to look at position  248 with base  T\n",
      "Current index is  249  whith base  G\n",
      "Current index is  250  whith base  T\n",
      "Current index is  251  whith base  G\n",
      "Current index is  252  whith base  G\n",
      "Current index is  253  whith base  G\n",
      "Current index is  254  whith base  T\n",
      "Current index is  255  whith base  T\n",
      "Current index is  256  whith base  C\n",
      "Current index is  257  whith base  C\n",
      "Current index is  258  whith base  T\n",
      "Current index is  259  whith base  T\n",
      "Current index is  260  whith base  T\n",
      "Current index is  261  whith base  A\n",
      "\n",
      "A flanking on the right side of the mutations was found at index  261\n",
      "\n",
      "Currently looking for the last base to make an overlap flanking the mutation  A15H\n",
      "\n",
      "Currently looking for the A to make an overlap flanking the left side of the mutation.\n",
      "The index range to look for the last overlap base is range(231, 226, -1)\n",
      "Current base is T\n",
      "Current base is C\n",
      "Current base is C\n",
      "Current base is G\n",
      "Current base is T\n",
      "\n",
      "Currently looking for the T to make an overlap flanking the left side of the mutation.\n",
      "The index range to look for the last overlap base is range(266, 271)\n",
      "Current base is T\n",
      "Found end position at 266\n",
      "\n",
      "Currently looking for the first base to make an overlap flanking the mutation  A15H\n",
      "Starting to look at position  235 with base  C\n",
      "Current index is  234  whith base  G\n",
      "Current index is  233  whith base  G\n",
      "Current index is  232  whith base  C\n",
      "Current index is  231  whith base  T\n",
      "\n",
      "T flanking on the left side of the mutations was found at index  231\n",
      "Starting to look at position  248 with base  T\n",
      "Current index is  249  whith base  G\n",
      "Current index is  250  whith base  T\n",
      "Current index is  251  whith base  G\n",
      "Current index is  252  whith base  G\n",
      "Current index is  253  whith base  G\n",
      "Current index is  254  whith base  T\n",
      "Current index is  255  whith base  T\n",
      "Current index is  256  whith base  C\n",
      "Current index is  257  whith base  C\n",
      "Current index is  258  whith base  T\n",
      "Current index is  259  whith base  T\n",
      "Current index is  260  whith base  T\n",
      "Current index is  261  whith base  A\n",
      "\n",
      "A flanking on the right side of the mutations was found at index  261\n",
      "\n",
      "Currently looking for the last base to make an overlap flanking the mutation  A15H\n",
      "\n",
      "Currently looking for the A to make an overlap flanking the left side of the mutation.\n",
      "The index range to look for the last overlap base is range(226, 221, -1)\n",
      "Current base is T\n",
      "Current base is A\n",
      "Found end position at 225\n",
      "\n",
      "Currently looking for the T to make an overlap flanking the left side of the mutation.\n",
      "The index range to look for the last overlap base is range(266, 271)\n",
      "Current base is T\n",
      "Found end position at 266\n",
      "\n",
      "Making the overlapping sequence for the mutation  A15H\n",
      "Left forward overlapping sequence is:  ATTGCCT\n",
      "Left reverse overlapping sequence is:  AGGCAAT\n",
      "Right forward overlapping sequence is:  AACGGT\n",
      "Right reverse overlapping sequence is:  ACCGTT\n",
      "\n",
      "Making the non-overlapping sequence for the mutation  A15H\n",
      "\n",
      "Currently calculating the non-overlapping region for the primer.\n",
      "\n",
      "Start search position is:  248 with base T\n",
      "\n",
      "Current position is:  258\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  259\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  260\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  261\n",
      "Base in this position is:  A\n",
      "\n",
      "Current position is:  262\n",
      "Base in this position is:  A\n",
      "\n",
      "Current position is:  263\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  264\n",
      "Base in this position is:  G\n",
      "\n",
      "Current position is:  265\n",
      "Base in this position is:  G\n",
      "\n",
      "Current position is:  266\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  267\n",
      "Base in this position is:  G\n",
      "\n",
      "Current position is:  268\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  269\n",
      "Base in this position is:  G\n",
      "\n",
      "The non-overlapping sequence of the primer of this insert is TGTGGGTTCCTTTAACGGTGCG\n",
      "and the Tm is 57.7124353414016\n",
      "\n",
      "Currently calculating the non-overlapping region for the primer.\n",
      "\n",
      "Start search position is:  224 with base G\n",
      "\n",
      "Current position is:  214\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  213\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  212\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  211\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  210\n",
      "Base in this position is:  A\n",
      "\n",
      "Current position is:  209\n",
      "Base in this position is:  A\n",
      "\n",
      "Current position is:  208\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  207\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  206\n",
      "Base in this position is:  G\n",
      "\n",
      "Current position is:  205\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  204\n",
      "Base in this position is:  A\n",
      "\n",
      "Current position is:  203\n",
      "Base in this position is:  g\n",
      "\n",
      "Current position is:  202\n",
      "Base in this position is:  t\n",
      "\n",
      "Current position is:  201\n",
      "Base in this position is:  a\n",
      "\n",
      "Current position is:  200\n",
      "Base in this position is:  t\n",
      "\n",
      "Current position is:  199\n",
      "Base in this position is:  a\n",
      "\n",
      "Current position is:  198\n",
      "Base in this position is:  c\n",
      "\n",
      "The non-overlapping sequence of the primer of this insert is catatgATGTCAACTCCTTCGATCGTG\n",
      "and the Tm is 56.05645960154874\n",
      "\n",
      "Currently calculating the non-overlapping region for the primer.\n",
      "\n",
      "Start search position is:  267 with base G\n",
      "\n",
      "Current position is:  277\n",
      "Base in this position is:  A\n",
      "\n",
      "Current position is:  278\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  279\n",
      "Base in this position is:  A\n",
      "\n",
      "Current position is:  280\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  281\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  282\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  283\n",
      "Base in this position is:  C\n",
      "\n",
      "The non-overlapping sequence of the primer of this insert is GCGTTCGCCAACACCCC\n",
      "and the Tm is 56.2303123535965\n",
      "\n",
      "Currently calculating the non-overlapping region for the primer.\n",
      "\n",
      "Start search position is:  245 with base G\n",
      "\n",
      "Current position is:  235\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  234\n",
      "Base in this position is:  G\n",
      "\n",
      "Current position is:  233\n",
      "Base in this position is:  G\n",
      "\n",
      "Current position is:  232\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  231\n",
      "Base in this position is:  T\n",
      "\n",
      "Current position is:  230\n",
      "Base in this position is:  C\n",
      "\n",
      "Current position is:  229\n",
      "Base in this position is:  C\n",
      "\n",
      "The non-overlapping sequence of the primer of this insert is CCTCGGCTGCTCGGACG\n",
      "and the Tm is 57.12078255679654\n",
      "\n",
      "Making the inter-sequence for the mutation  A15H\n",
      "Left forward primer inter-sequence is  CGGCTGCTCGGACGCA\n",
      "Right reverse primer inter-sequence is  AAAGGAACCCACATG\n",
      "\n",
      "Replacing T with U in the overlapping region and combining with the non-overlapping region to finalize the primers for the variant  A15H\n",
      "The final left forward primer is  ATTGCCUCGGCTGCTCGGACGCATGTGGGTTCCTTTAACGGTGCG\n",
      "The final left reverse primer is  AGGCAAUCACGATCGAAGGAGTTGACATcatatg\n",
      "The final right forward primer is  AACGGUGCGTTCGCCAACACCCC\n",
      "The final right reverse primer is  ACCGTUAAAGGAACCCACATGCGTCCGAGCAGCCGAGG\n",
      "\n",
      "Calculating GC content of designed primers.\n",
      "\n",
      "GC content of primer_left_fw is  61.36363636363637\n",
      "\n",
      "GC content of primer_left_rv is  45.45454545454545\n",
      "\n",
      "GC content of primer_right_fw is  68.18181818181817\n",
      "\n",
      "GC content of primer_right_rv is  62.16216216216216\n",
      "Homopolymeric penalty for this primer is  0\n",
      "Homopolymeric penalty for this primer is  0\n",
      "Homopolymeric penalty for this primer is  0\n",
      "Homopolymeric penalty for this primer is  0\n",
      "Total score left is  94.9090909090909\n",
      "Total score right is  91.34398034398033\n"
     ]
    }
   ],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
