{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "690e282b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils import MeltingTemp\n",
    "import pandas as pd\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "58810460",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify the Excel file path and sheet name\n",
    "excel_file = \"codon_usage_frequency_escherichia_coli_b.xlsx\"\n",
    "\n",
    "# Read the Excel file into a DataFrame\n",
    "df = pd.read_excel(excel_file)\n",
    "\n",
    "# Create a dictionary from the Excel columns\n",
    "amino_acids = df[\"Amino acid\"].tolist()\n",
    "codons_1 = df.groupby(\"Amino acid\")[\"Codon 1\"].apply(list).to_dict()\n",
    "codons_2 = df.groupby(\"Amino acid\")[\"Codon 2\"].apply(list).to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5a8dcf87",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('plasmid.txt', 'r') as plasmid_file:\n",
    "    plasmid_lines = plasmid_file.readlines()\n",
    "\n",
    "plasmid_seq = Seq(plasmid_lines[0].strip())\n",
    "start_gene_index = int(plasmid_lines[1]) - 1\n",
    "end_gene_index = int(plasmid_lines[2]) - 1\n",
    "gene_seq = plasmid_seq[start_gene_index:(end_gene_index + 1)]\n",
    "protein_seq = gene_seq.translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ba7e0327",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify the path to your .txt file\n",
    "mutations_path = 'mutations.txt'\n",
    "\n",
    "# Create an empty list to store the values\n",
    "mutations_list = []\n",
    "\n",
    "# Count the number of mutations\n",
    "mutations_count = 0\n",
    "\n",
    "# Open the file for reading\n",
    "with open(mutations_path, 'r') as mutations_file:\n",
    "    # Read each line and append it to the list\n",
    "    for line in mutations_file:\n",
    "        mutations_list.append(line.strip())  # strip() removes any leading/trailing whitespace\n",
    "        mutations_count += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "36c56047",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mutation now is  T14A\n",
      "Mutation now is  N25K\n",
      "Mutation now is  S37V\n"
     ]
    }
   ],
   "source": [
    "# Identify and separate components in mutations\n",
    "\n",
    "mutations_separated = [[],[],[]]\n",
    "\n",
    "def separate_mutation(mutation) :\n",
    "    \n",
    "    print('Mutation now is ', mutation)\n",
    "\n",
    "    # Define a regular expression pattern to match a letter followed by a number and then another letter\n",
    "    pattern = r'([A-Za-z])(\\d{1,4})([A-Za-z])'\n",
    "\n",
    "    # Use the search function to find the match in the input string\n",
    "    match = re.search(pattern, mutation)\n",
    "\n",
    "    # Check if a match was found\n",
    "    if match:\n",
    "        # Extract the individual components into variables\n",
    "        original_AA = match.group(1)\n",
    "        position = match.group(2)\n",
    "        new_AA = match.group(3)\n",
    "    else:\n",
    "        print(\"Could not find correct mutation pattern.\")\n",
    "\n",
    "    return original_AA, position, new_AA\n",
    "\n",
    "for mutation in mutations_list :\n",
    "    separated_mutation = separate_mutation(mutation)\n",
    "    mutations_separated[0].append(separated_mutation[0])\n",
    "    mutations_separated[1].append(int(separated_mutation[1]))\n",
    "    mutations_separated[2].append(separated_mutation[2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ef8b1555",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original codon is  ACG\n",
      "Original codon is  AAC\n",
      "Original codon is  TCT\n"
     ]
    }
   ],
   "source": [
    "# Find position of mutation in the gene:\n",
    "mutations_position_gene = []\n",
    "original_codons = []\n",
    "new_codons = []\n",
    "new_plasmid_seqs = []\n",
    "\n",
    "for mutation in range(mutations_count) :\n",
    "    codon_start_gene = ((mutations_separated[1][mutation] - 1) * 3)\n",
    "    mutations_position_gene.append(codon_start_gene)\n",
    "    codon_end_gene = codon_start_gene + 3\n",
    "    original_codon = gene_seq[codon_start_gene:codon_end_gene]\n",
    "    print('Original codon is ', original_codon)\n",
    "    original_codons.append(original_codon)\n",
    "    new_codon = Seq(str(codons_1[mutations_separated[2][mutation]]))\n",
    "    print('New codon is ', new_codon)\n",
    "    new_codons.append(new_codon)\n",
    "    new_plasmid_seq = plasmid_seq[:(start_gene_index + codon_start_gene)] + new_codon + plasmid_seq[(start_gene_index + codon_end_gene):]\n",
    "    new_plasmid_seqs.append(new_plasmid_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a61ea0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_overlap_flanking_position(start_insert_index, plasmid, id, overlap_flanking_condition, overlap_end_search_direction):\n",
    "    print('\\n\\nCurrently looking for the correct base flanking the insert ID', id)\n",
    "    print('Overlap direction is', overlap_end_search_direction)\n",
    "    print('Start looking at position', start_insert_index)\n",
    "    print('And base is', plasmid[start_insert_index])\n",
    "\n",
    "    search_range_start_overlap = [overlap_end_search_direction, 2 * overlap_end_search_direction, 0]\n",
    "\n",
    "    for position_change in search_range_start_overlap:\n",
    "        position = start_insert_index + position_change\n",
    "        print('Current base is', plasmid[position])\n",
    "        if plasmid[position] in overlap_flanking_condition:\n",
    "            overlap_flanking_position = position\n",
    "            print('Found start position at', overlap_flanking_position)\n",
    "            break\n",
    "    else:\n",
    "        print('\\nError: could not find the correct base flanking the insert ID', id)\n",
    "        overlap_flanking_position = None\n",
    "\n",
    "    return overlap_flanking_position\n",
    "\n",
    "\n",
    "def find_overlap_end_position(overlap_flanking_position, plasmid, id, overlap_end_condition, overlap_end_search_direction):\n",
    "    print('\\nCurrently looking for the correct base to make an overlap flanking the insert ID', id)\n",
    "    print('Overlap direction is', overlap_end_search_direction)\n",
    "    print('Start looking at position', overlap_flanking_position)\n",
    "    print('And base is', plasmid[overlap_flanking_position])\n",
    "\n",
    "    search_range_end_overlap = range((overlap_flanking_position + (5 * overlap_end_search_direction)),\n",
    "                                     (overlap_flanking_position + (10 * overlap_end_search_direction)),\n",
    "                                     overlap_end_search_direction)\n",
    "\n",
    "    print('The range to look for the last overlap base is', search_range_end_overlap)\n",
    "\n",
    "    for position in search_range_end_overlap:\n",
    "        print('Current base is', plasmid[position])\n",
    "        if plasmid[position] in overlap_end_condition:\n",
    "            overlap_end_position = position\n",
    "            print('Found end position at', overlap_end_position)\n",
    "            break\n",
    "    else:\n",
    "        print('\\nError: could not find the correct base to make an overlap flanking the insert ID', id)\n",
    "        overlap_end_position = None\n",
    "\n",
    "    return overlap_end_position\n",
    "\n",
    "\n",
    "def calculate_non_overlap(plasmid, last_overlap_position, id, non_overlap_search_range, primer_direction):\n",
    "    print('\\nCurrently calculating the length of the non-overlapping region for the primer of the insert ID', id)\n",
    "\n",
    "    start_position = last_overlap_position + primer_direction\n",
    "    print('\\nStart search position is:', start_position, 'with base', plasmid[start_position])\n",
    "\n",
    "    for non_overlap_length in non_overlap_search_range:\n",
    "        current_position = start_position + non_overlap_length\n",
    "        print('\\nCurrent position is:', current_position)\n",
    "        print('Base in this position is:', plasmid[current_position])\n",
    "\n",
    "        if current_position > start_position:\n",
    "            seq = plasmid[start_position:(current_position + 1)]\n",
    "        else:\n",
    "            seq = plasmid[current_position:(start_position + 1)]\n",
    "\n",
    "        tm = MeltingTemp.Tm_NN(seq)\n",
    "\n",
    "        if (tm >= 56 and seq[-1] in ['C', 'c', 'G', 'g']) or (tm >= 56 and seq[0] in ['C', 'c', 'G', 'g']):\n",
    "            break\n",
    "\n",
    "    else:\n",
    "        print('\\nError: could not resolve non-overlapping Tm for the primer of the insert ID', id)\n",
    "        seq = None\n",
    "        tm = None\n",
    "\n",
    "    print('\\nThe non-overlapping sequence of the primer of this insert is', seq)\n",
    "\n",
    "    return seq, tm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff902e6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def primer_calculator(id, plasmid, start_insert_index, overlap_end_search_direction):\n",
    "    if overlap_end_search_direction == -1:\n",
    "        overlap_flanking_condition = ['T', 't']\n",
    "        overlap_end_condition = ['A', 'a']\n",
    "    elif overlap_end_search_direction == 1:\n",
    "        overlap_flanking_condition = ['A', 'a']\n",
    "        overlap_end_condition = ['T', 't']\n",
    "    else:\n",
    "        overlap_flanking_condition = None\n",
    "        overlap_end_condition = None\n",
    "        print('Wrong value for overlap end search direction.')\n",
    "\n",
    "    overlap_flanking_position = find_overlap_flanking_position(start_insert_index, plasmid, id,\n",
    "                                                               overlap_flanking_condition, overlap_end_search_direction)\n",
    "\n",
    "    overlap_end_position = find_overlap_end_position(overlap_flanking_position, plasmid, id,\n",
    "                                                     overlap_end_condition, overlap_end_search_direction)\n",
    "\n",
    "    print('\\nMaking the overlapping sequence for the insert ID', id)\n",
    "\n",
    "    if overlap_end_search_direction == -1:\n",
    "        overlap_seq = plasmid[overlap_end_position:(overlap_flanking_position + 1)]\n",
    "    if overlap_end_search_direction == 1:\n",
    "        overlap_seq = plasmid[overlap_flanking_position:(overlap_end_position + 1)]\n",
    "    overlap_seq_rv = overlap_seq.reverse_complement()\n",
    "\n",
    "    print('Overlapping sequence is:', overlap_seq)\n",
    "    print('Reverse overlapping sequence is:', overlap_seq_rv)\n",
    "\n",
    "    non_overlap_search_range_fwd = range(10, 60)\n",
    "    non_overlap_search_range_rv = range(-10, -60, -1)\n",
    "\n",
    "    if overlap_end_search_direction == -1:\n",
    "        non_overlap_fw = calculate_non_overlap(plasmid, overlap_flanking_position, id, non_overlap_search_range_fwd, 1)\n",
    "        non_overlap_fw_seq = non_overlap_fw[0]\n",
    "        non_overlap_fw_tm = non_overlap_fw[1]\n",
    "        non_overlap_rv = calculate_non_overlap(plasmid, overlap_end_position, id, non_overlap_search_range_rv, -1)\n",
    "        non_overlap_rv_seq = non_overlap_rv[0].reverse_complement()\n",
    "        non_overlap_rv_tm = non_overlap_rv[1]\n",
    "    elif overlap_end_search_direction == 1:\n",
    "        non_overlap_fw = calculate_non_overlap(plasmid, overlap_end_position, id, non_overlap_search_range_fwd, 1)\n",
    "        non_overlap_fw_seq = non_overlap_fw[0]\n",
    "        non_overlap_fw_tm = non_overlap_fw[1]\n",
    "        non_overlap_rv = calculate_non_overlap(plasmid, overlap_flanking_position, id, non_overlap_search_range_rv, -1)\n",
    "        non_overlap_rv_seq = non_overlap_rv[0].reverse_complement()\n",
    "        non_overlap_rv_tm = non_overlap_rv[1]\n",
    "    else:\n",
    "        print('Wrong value for overlap end search direction.')\n",
    "\n",
    "    print('\\nReplacing T with U in the overlapping region and combining with the non-overlapping region to finalize the primer for the insert ID', id)\n",
    "    primer_fw = overlap_seq[:-1] + 'U' + non_overlap_fw_seq\n",
    "    print('The final primer is', primer_fw)\n",
    "    primer_rv = overlap_seq_rv[:-1] + 'U' + non_overlap_rv_seq\n",
    "    print('The final reverse primer is', primer_rv)\n",
    "\n",
    "    return primer_fw, non_overlap_fw_tm, primer_rv, non_overlap_rv_tm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "887db3d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def primers_insert_calculator(mutations_list, new_plasmid_seqs, start_insert_index, overlap_end_search_direction):\n",
    "    primers_list = []\n",
    "\n",
    "    for mutation_number in range(mutations_count):\n",
    "        start_point = start_insert_index\n",
    "        if overlap_end_search_direction == 1:\n",
    "            start_point += (insert_lengths[mutation_number] - 1)\n",
    "        primer_fw, non_overlap_fw_tm, primer_rv, non_overlap_rv_tm = primer_calculator(mutations_list[mutation_number], new_plasmid_seqs[mutation_number], start_point, overlap_end_search_direction)\n",
    "        primers_list.append([primer_fw, non_overlap_fw_tm, primer_rv, non_overlap_rv_tm])\n",
    "\n",
    "    return primers_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83ab2145",
   "metadata": {},
   "outputs": [],
   "source": [
    "overlap_end_search_direction_5_line = -1\n",
    "results_5_line = primers_insert_calculator(plasmid_ids, plasmids, start_insert_index, overlap_end_search_direction_5_line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5e34088",
   "metadata": {},
   "outputs": [],
   "source": [
    "transposed_results_5_line = [[results_5_line[j][i] for j in range(number_seqs)] for i in range(len(results_5_line[0]))]\n",
    "\n",
    "final_primers_seq_list = []\n",
    "final_primers_names_list = []\n",
    "final_primers_number_list = []\n",
    "final_primers_tm_list = []\n",
    "\n",
    "for number in range(number_seqs):\n",
    "    final_primers_seq_list.append(transposed_results_5_line[0][number])\n",
    "    final_primers_seq_list.append(transposed_results_5_line[2][number])\n",
    "\n",
    "    final_primers_names_list.append(plasmid_ids[number] + ' insert fw')\n",
    "    final_primers_names_list.append(plasmid_ids[number] + ' insert rv')\n",
    "\n",
    "    final_primers_tm_list.append(transposed_results_5_line[1][number])\n",
    "    final_primers_tm_list.append(transposed_results_5_line[3][number])\n",
    "\n",
    "    for n in range(1, 3):\n",
    "        final_primers_number_list.append(n)\n",
    "\n",
    "df = pd.DataFrame({'Name': final_primers_names_list, 'Number': final_primers_number_list, 'Short name': None, 'Sequence': final_primers_seq_list, 'Tm': final_primers_tm_list})\n",
    "df.to_excel(\"primers_list.xlsx\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
